fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/WorldRenderContext.java

```diff
@@ -0,0 +1,203 @@
+	 *
+	 * @return {@code true} when "fabulous" graphics mode is enabled.
+	 */
+	boolean advancedTranslucency();
+
+	public interface Drawable {
+		/**
+		 * The {@code VertexConsumerProvider} instance being used by the world renderer for most non-terrain renders.
+		 * Generally this will be better for most use cases because quads for the same layer can be buffered
+		 * incrementally and then drawn all at once by the world renderer.
+		 *
+		 * <p>IMPORTANT - all vertex coordinates sent to consumers should be relative to the camera to
+		 * be consistent with other quads emitted by the world renderer and other mods.  If this isn't
+		 * possible, caller should use a separate "immediate" instance.
+		 *
+		 * <p>This property is not exposed before {@link WorldRenderEvents#BEFORE_ENTITIES} or after
+		 * {@link WorldRenderEvents#BEFORE_DEBUG_RENDER} because the consumer buffers are not available before or
+		 * drawn after that in vanilla world rendering.  Renders that cannot draw in one of the supported events
+		 * must be drawn directly to the frame buffer, preferably in {@link WorldRenderEvents#LAST} to avoid being
+		 * overdrawn or cleared.
+		 */
+		VertexConsumerProvider consumers();
+	}
+
+	/**
+	 * Conveys WorldRenderer state to event handlers after invocation.
+	 * Includes view frustum, which is not available then.
+	 */
+	public interface AfterFrustum {
+		Frustum frustum();
+	}
+
+	/**
+	 * Used in the block breaking callback to convey information on
+	 * ray trace hit and accept cancellation of default outline render.
+	 */
+	public interface PreBlockOutline {
+		/**
+		 * The game object currently under the crosshair target. Normally
+		 * equivalent to {@link MinecraftClient#crosshairTarget}.
+		 * Provided for convenience.
+		 */
+		@Nullable HitResult hitResult();
+
+		/**
+		 * Call to prevent the default block outline from rendering.
+		 * Use this for event handlers that fully replace the vanilla
+		 * block outline render.
+		 *
+		 * <p>The effect of this method depends on the event from which it is called:
+		 *
+		 * <p>In {@code BEFORE_BLOCK_RENDER} this prevents invocation of
+		 * the {@code BLOCK_RENDER} event and {@link AfterBlockOutline#didCancelDefaultBlockOutline()}
+		 * will be {@code false} in later-stage event invocations. It (obviously)
+		 * also prevents the default outline render from running.
+		 *
+		 * <p>In {@code BLOCK_RENDER} this prevents the default outline render
+		 * from running and {@link AfterBlockOutline#didCancelDefaultBlockOutline()}
+		 * will be {@code true} in later-stage event invocations.
+		 *
+		 * <p>In both cases, this method is idempotent - calling it from more
+		 * than one event subscriber has the same effect as calling it once.
+		 * All event handlers within the same event invocation will always
+		 * execute and the cancellation status cannot be inspected during the
+		 * block outline render events.
+		 */
+		void cancelDefaultBlockOutline();
```

Comment:
Player raised a concern here about this context in use mixing up the cancellation + notification handling.

Reply:
I'll need a little more elaboration to understand what the problem is here.

Reply:
We want to avoid mixing notification and cancellation in the same event since that'd incorrectly notify some subscribers. If the 2nd subscriber happened to cancel, the 1st one would never know that the notified action won't actually happen with inline cancellation.

So e.g. instead of
```
foreach (subscriber) {
  if (!subscriber.onBeforeX()) return;
}

<vanilla logic>

foreach (subscriber) {
  subscriber.onAfterX();
}
```
it should do
```
boolean cancelled = false;

foreach (subscriber) {
  if (!subscriber.shouldDoX()) {
    cancelled = true;
    break;
  }
}

if (!cancelled) {
  foreach (subscriber) {
    subscriber.onBeforeX();
  }

  <vanilla logic>

  foreach (subscriber) {
    subscriber.onAfterX();
  }
} else {
  foreach (subscriber) {
    subscriber.onXCancelled();
  }
}
```
for an example with the full set of notifications (a real case would only provide those that are reasonable)

Reply:
That helps, thank you.  I'll need to think about how it applies here.

Reply:
I've considered this some more.  The main problem seems to be the API causes confusion.

First, the behavior is not like this:

> We want to avoid mixing notification and cancellation in the same event since that'd incorrectly notify some subscribers. If the 2nd subscriber happened to cancel, the 1st one would never know that the notified action won't actually happen with inline cancellation.

The block render events cannot cancel an event for any peer subscriber.  Ever.  This is clearly stated in the docs:

> ... [cancelDefaultBlockOutline()] is idempotent - calling it from more than one event subscriber has the same effect as calling it once. **All event handlers within the same event invocation will always execute** and the cancellation status cannot be inspected during the block outline render events.

The "cannot be inspected" is specifically meant to prohibit subscribers from trying to compete with or depend on each other.  The order of calling is not deterministic and they don't directly affect each other, so that would only cause problems.

What it cancels depends on usage. 
* In `BEFORE_BLOCK_RENDER` it prevents the block outline render entirely. `BLOCK_RENDER` will not be called.
* In `BLOCK_RENDER` it prevents the *vanilla* block outline render from running but has no effect on other subscribers.

Let's cover each in more detail.

**`BEFORE_BLOCK_RENDER`**
The primary use here is when the player is looking at something like a fluid or entity that wants to mimic block behavior and display an outline. A fluid or entity would never register as a hit for block outline rendering, so this provides a way to detect those special hits and render if appropriate.  

If the block outline render isn't cancelled here, then depending on circumstance _two_ outlines will be rendered.   The desired behavior is to pretend there was never a block hit in the first place, and so the entire block outline handling is cancelled and `BLOCK_RENDER` never fires, which is exactly what should normally happen if the player isn't targeting a block.

The `BEFORE_BLOCK_RENDER` is also potentially useful for rendering in-world information that doesn't necessarily care about whether or not block outline rendering occurs.  Subscribers with this use case can rely on being called every frame even if the block outline render is cancelled.

What happens if an entity with a special bounding box is standing in a fluid that also has a special bounding box?   Both boxes will render. Any other answer would make this much more elaborate than it needs to be for the vast majority of cases. 

**`BLOCK_RENDER`**
This is called when a conventional block outline render would normally happen - when the player is targeting a nearby block and we aren't pretending otherwise because some mod is faking a hit.

This is where mods that want to replace block outline rendering for specific blocks should subscribe.  Every subscriber gets called, and every mod that wants to can render.  Exotic Blocks, for example, has many non-cubic shapes that would not render well with a conventional outline renderer so it can use this event instead of a mixin to do that.

Again, the subscribers here cannot cancel each other - they can only cancel the default block outline render that would normally come after.

This event is *not* a good solution for mods that want to replace the default outline renderer for *all* blocks.  First, such a mod should probably not render an outline if a subscriber to `BLOCK_RENDER` has cancelled it.  Presumably a more specific outline renderer will do a better job and should be allowed to do it.

I assumed a general-purpose outline mod would not use the `BLOCK_RENDER` event and instead mix in to `drawShapeOutline`, which will not be called if `BLOCK_RENDER` is cancelled. That's a nice, clean target and generally only one mod is going to be able to replace it fully.

However, at least one such mod (Custom Selection Box) doesn't want to render at that point because it is going to employ translucency that wouldn't render correctly until later in the frame.  It is going to `NOOP` the `drawShapeOutline` and render the outline in a later WorldRenderEvent. 

However, such a mod needs a way to know if `drawShapeOutline` was ever called.  It *could* set some state there and then reference it a few microseconds later, but for convenience I added `didCancelDefaultBlockOutline()` to the context so it can check directly.  But that's a false economy because said mod still needs to mix in to `drawShapeOutline` to prevent it from doing anything.

And the larger problem is that the overall scheme around block outline rendering is very confusing.

I think it may be best to simplify the two block outline events and remove `didCancelDefaultBlockOutline()`.  General-purpose block outline replacements can fend for themselves using the usual and customary methods. More robust support for a general-purpose outline replacement would require another event, at least one more mixin, or perhaps some kind of `BlockOutlineRendererPlugIn` API, all of which seems well outside the scope of this PR, and probably outside the scope of Fabric API generally.